src/abstract/CSBondCurve.sol

```diff
@@ -95,98 +97,177 @@ abstract contract CSBondCurve is ICSBondCurve, Initializable {
+        }
     }
 
     function _getKeysCountByBondAmount(
         uint256 amount,
-        BondCurve storage curve
+        BondCurveInterval[] storage intervals
     ) internal view returns (uint256) {
-        if (amount < curve.points.unsafeAccess(0).value) {
+        if (amount < intervals[0].fromBond) {
             return 0;
         }
 
-        uint256 len = curve.points.length;
-        uint256 maxCurveAmount = curve.points.unsafeAccess(len - 1).value;
         unchecked {
+            uint256 low = 0;
+            uint256 high = intervals.length - 1;
+            while (low < high) {
+                uint256 mid = (low + high + 1) / 2;
+                BondCurveInterval storage midInterval = intervals[mid];
+                if (amount <= midInterval.fromBond) {
+                    if (amount == midInterval.fromBond) {
+                        return intervals[mid].fromKeysCount;
+                    }
+                    high = mid - 1;
+                } else {
+                    low = mid;
+                }
+            }
+            BondCurveInterval storage interval = intervals[low];
             return
-                amount < maxCurveAmount
-                    ? _searchKeysCount(amount, curve.points)
-                    : len + (amount - maxCurveAmount) / curve.trend;
+                interval.fromKeysCount +
+                (amount - interval.fromBond) /
+                interval.trend;
         }
     }
 
     // solhint-disable-next-line func-name-mixedcase
     function __CSBondCurve_init(
-        uint256[] calldata defaultBondCurvePoints
+        uint256[2][] calldata defaultBondCurveIntervals
     ) internal onlyInitializing {
-        uint256 addedId = _addBondCurve(defaultBondCurvePoints);
+        uint256 addedId = _addBondCurve(defaultBondCurveIntervals);
         if (addedId != DEFAULT_BOND_CURVE_ID) {
             revert InvalidInitialisationCurveId();
         }
     }
 
+    /// @dev TODO: Remove the method in the next major release.
+    ///      Migrate legacy bond curves to the new format.
+    ///      It should be called only once during the upgrade to CSM v2.
+    function __addBondCurvesWithIntervals() internal onlyInitializing {
+        // TODO: Check values after upgrade in tests
+        uint256[2][] memory defaultBondCurve = new uint256[2][](2);
+        defaultBondCurve[0] = [uint256(1), 2.4 ether];
+        defaultBondCurve[1] = [uint256(2), 1.3 ether];
+        _addBondCurveOnMigration(defaultBondCurve);
+
+        uint256[2][] memory vettedGateBondCurve = new uint256[2][](2);
+        vettedGateBondCurve[0] = [uint256(1), 1.5 ether];
+        vettedGateBondCurve[1] = [uint256(2), 1.3 ether];
+        _addBondCurveOnMigration(vettedGateBondCurve);
+    }
+
+    /// @dev TODO: Remove the method in the next major release.
+    ///      Helper function for migrating legacy bond curves to the new format.
+    function _addBondCurveOnMigration(uint256[2][] memory intervals) internal {
```

Comment:
We can just use add bond curve method I guess

Reply:
We can't pass `uint256[][] memory` as `uint256[][] calldata`, unfortunately